### DISCUSSION TITLE
ESP32-DEV pwm controlling 3v LEDs without MOSFET

### DISCUSSION THREAD

**Discussion by rubenkr** [arendst/Tasmota#18727] [Q&A]
Hi everyone,

I am frustrating here with a so simple task that I am afraid to ask. But I need to solve it. Kindly asking for your help.

Having a fairy light mit 20 LEDs powered with 3V I wanted to put a ESP in between to control brightness.

Reading that ESP32 can control PWM best I opted for a one that I used in another LED project the QuinLED which is ESP32-DEVKIT. In Tasmota I set the GPIO 21 as PWM and connect the LED strip - to GPIO 21 on ESP and + to 3,3V Pin on ESP.

But even on full brightness I see only a fraction of the full brightness. When I connect the 3V Pin and GND to the LED directly it runs at 100% brightness. However I am experiencing wifi drops. 

What am I doing wrong? How in hell can I solve this problem with being tasmota controllable?

THANK YOU

Picture: https://imgur.com/a/t37OzWf

**Comment by joba-1** [comment]
Hi, that is easy (if you know it ;P ): 

Microcontrollers like the ESP32 cannot provide much current via its pins. Usually it is just enough for one led (~20mA).

**Reply by joba-1** [reply]
And before you ask: you could use more than one pin in parallel and it probably gets a little brighter. But there is also a tight current limit for all pins summed. 
And then the 3.3V of most ESP32 boards cannot provide much extra current. So even a mosfet driven by this power might not work (makes the ESP32 crash frequently).

**Reply by sfromis** [reply]
The best you could get from ESP32 would be one led per gpio, and up to 16 PWM pins. If the leds are not too bright, you could likely get 2-3 leds per pin, but as joba-1 pointed out, you cannot just add up and assume that you can get the max power from all pins.

If you want "many" leds, without a MOSFET driver, you could use a neopixel addressable led strip (WS2812 or similar), where the MCU only has to provide a control signal,with the strip being wired directly to both power (like 5V) and ground. Such a strip can be used without dealing with pixels separately, but act as a single light, if you prefer.

**Comment by rubenkr** [comment]
Thank you all very much. At least there was not other option. I tried using a mosfet gate on gpiox drain to gnd led and source to gnd on esp but experienced heavy Wi-Fi drops.

Is there any off the shelf chip that I can buy with better wifi and mosfet? I found the Sinlink MOS but it only outputs 5V and not dimmer functionality.

Your ideas are much appreciated

**Reply by joba-1** [reply]
Since you already tried a mosfet: did you switch the 3.3V or the 5V?
Switching the 5V would be worth a try!
Don’t forget the resistor to protect the leds.

**Reply by rubenkr** [reply]
Hi! I switched the 3V pin but i think the 5V would not work as the led to not ignite when hooked up to 5V?

**Reply by joba-1** [reply]
that is what the resistor in line with the LEDs is good for.
For one typical LED calculate like this: the LED „uses“ about 3V when fully on. So the resistor has to use the other 2V. The LED can take roughly up to 20mA (0.02A) current that also flows through the resistor. With Ohms law you get the resistor value R = 2V/0.02A = 100 Ohm.
Of course your LEDs might have different values. You should start with 200 Ohm and reduce slowly until you get the expected brightness.

**Reply by rubenkr** [reply]
@joba-1 that is a awesome idea for the out-of-the-box chip from Sinlink MOS. I solved it yesterday with an ESP32-DEV using 3V pin and MOSFET. Wifi is stable now. Maybe the last ESP8266 had a flaw.

Two questions:
1. How to measure the mA used by a LED with the multimeter?
2. If LEDs are in line (e.g. 30 pieces like a fairylight) is the calculation the same (e.g. 30x200Ohm Resistor on the power input)?

I learned new things of you and will try it next project with the Sinlink MOS. Already crawling my head how to solder a resistor in line with the output of the [Sinilink ](https://templates.blakadder.com/sinilink_XY-WFMS.html)as it comes presoldered. But this will be a another topic. Much thanks <3

**Reply by joba-1** [reply]
1) Open the circuit at one point and close it again by connecting the tips of the multimeter with the two ends (current is the same in the whole circuit)

2) The leds are not in line, they are all in parallel or you would need 90V to turn 30 leds on.
That means the resistor is the same but the current will add up, e.g. 30*20mA = 0.6A

The voltage regulator on the sinilink looks beefy enough to provide this current, but that’s just a guess, haven’t checked the datasheet.

P.S. with 5V the resistor „burns“ 1.2W (2V*0.6A). Normal resistors cannot do that.

**Reply by sfromis** [reply]
Looks like that board has the pretty common AMS1117 voltage regulator which typically is 500 mA, meaning that drawing 0.6A may be "optimistic". However my *guess* would be that such a fairy light strip has smaller leds than 20 mA, making it plausible to draw the current from the voltage regulator.

**Reply by joba-1** [reply]
hm, looked closer, it is probably the kia1117 that can do 1A, 

I agree with the led current, but lower is not a problem:)

**Comment by rubenkr** [comment]
I would love to give it a real test try. @sfromis is right the led are really low power led lights. But I have not measured.

When I understood @sfromis correctly the Sinilink board has already a voltage regulator. How can I transform the 5V output terminals into 3V? Reading @joba-1 post I assume I have to hookup a resistor in line to the power output (e.g. one end of the 200Ohm resistor into VOUT and the other and to positive wire of the LED strip, right?). Sorry for my low skills on electronics.

**Reply by sfromis** [reply]
KIA1117 pinout shows 3.3V available on the middle pin: 
![image](https://github.com/arendst/Tasmota/assets/47082390/ff8986f3-311a-4b80-818c-16fe13963046)
